pkg/providers/gke.go

```diff
@@ -0,0 +1,342 @@
+/*
+Copyright 2018 The Shared LoadBalancer Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package providers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	kubeconv1alpha1 "github.com/Huang-Wei/shared-loadbalancer/pkg/apis/kubecon/v1alpha1"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/compute/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// GKE stands for IBM Kubernetes Service
+type GKE struct {
+	client *GCE
+	project, region string
+	// key is namespacedName of a LB Serivce, val is the service
+	cacheMap map[types.NamespacedName]*corev1.Service
+
+	// cr to LB is 1:1 mapping
+	crToLB map[types.NamespacedName]types.NamespacedName
+	// lb to CRD is 1:N mapping
+	lbToCRs map[types.NamespacedName]nameSet
+	// lbToPorts is keyed with ns/name of a LB, and valued with ports info it holds
+	lbToPorts map[types.NamespacedName]int32Set
+
+	capacityPerLB int
+}
+
+var _ LBProvider = &GKE{}
+var ctx = context.Background()
+
+type GCE struct {
+	*compute.Service
```

Comment:
where is `compute`? I guess in L27, "google.golang.org/api/compute/v1" should be aliased to "compute"?

Reply:
done